Bottom: 055ae4dba49ee991ba8a7b4c5daebc29c8ccb226
Top:    51046192c9d320027c538429400a7c9175dfdbab
Author: Peter Clifton <peter@clifton-electronics.co.uk>
Date:   2016-01-09 17:30:15 +0000

Crasy test


---

diff --git a/src/Makefile.am b/src/Makefile.am
index 717be1a..3b82b5e 100644
--- a/src/Makefile.am
+++ b/src/Makefile.am
@@ -58,6 +58,8 @@ PCB_SRCS = \
 	dolists.h \
 	draw.c \
 	draw.h \
+	draw_funcs.c \
+	draw_funcs.h \
 	drill.c \
 	drill.h \
 	edif.y \
diff --git a/src/draw.c b/src/draw.c
index 6a62f3a..8cc627e 100644
--- a/src/draw.c
+++ b/src/draw.c
@@ -40,6 +40,7 @@
 #include "crosshair.h"
 #include "data.h"
 #include "draw.h"
+#include "draw_funcs.h"
 #include "error.h"
 #include "mymem.h"
 #include "misc.h"
@@ -80,7 +81,6 @@ static HID_DRAW *hid_draw = NULL;
  * some local prototypes
  */
 static void DrawEverything (const BoxType *);
-static void DrawPPV (int group, const BoxType *);
 static void AddPart (void *);
 static void DrawEMark (ElementType *, Coord, Coord, bool);
 static void DrawRats (const BoxType *);
@@ -100,12 +100,6 @@ set_object_color (AnyObjectType *obj, char *warn_color, char *selected_color,
   hid_draw_set_color (Output.fgGC, color);
 }
 
-static void
-set_layer_object_color (LayerType *layer, AnyObjectType *obj)
-{
-  set_object_color (obj, NULL, layer->SelectedColor, PCB->ConnectedColor, PCB->FoundColor, layer->Color);
-}
-
 /*---------------------------------------------------------------------------
  *  Adds the update rect to the update region
  */
@@ -143,170 +137,6 @@ Redraw (void)
   gui->invalidate_all ();
 }
 
-static void
-_draw_pv_name (PinType *pv)
-{
-  BoxType box;
-  bool vert;
-  TextType text;
-
-  if (!pv->Name || !pv->Name[0])
-    text.TextString = EMPTY (pv->Number);
-  else
-    text.TextString = EMPTY (TEST_FLAG (SHOWNUMBERFLAG, PCB) ? pv->Number : pv->Name);
-
-  vert = TEST_FLAG (EDGE2FLAG, pv);
-
-  if (vert)
-    {
-      box.X1 = pv->X - pv->Thickness    / 2 + Settings.PinoutTextOffsetY;
-      box.Y1 = pv->Y - pv->DrillingHole / 2 - Settings.PinoutTextOffsetX;
-    }
-  else
-    {
-      box.X1 = pv->X + pv->DrillingHole / 2 + Settings.PinoutTextOffsetX;
-      box.Y1 = pv->Y - pv->Thickness    / 2 + Settings.PinoutTextOffsetY;
-    }
-
-  hid_draw_set_color (Output.fgGC, PCB->PinNameColor);
-
-  text.Flags = NoFlags ();
-  /* Set font height to approx 56% of pin thickness */
-  text.Scale = 56 * pv->Thickness / FONT_CAPHEIGHT;
-  text.X = box.X1;
-  text.Y = box.Y1;
-  text.Direction = vert ? 1 : 0;
-
-  if (hid_draw_is_gui (hid_draw))
-    doing_pinout++;
-  hid_draw_pcb_text (Output.fgGC, &text, 0);
-  if (hid_draw_is_gui (hid_draw))
-    doing_pinout--;
-}
-
-static void
-_draw_pv (PinType *pv, bool draw_hole)
-{
-  if (TEST_FLAG (THINDRAWFLAG, PCB))
-    hid_draw_thin_pcb_pv (Output.fgGC, Output.fgGC, pv, draw_hole, false);
-  else
-    hid_draw_fill_pcb_pv (Output.fgGC, Output.bgGC, pv, draw_hole, false);
-
-  if ((!TEST_FLAG (HOLEFLAG, pv) && TEST_FLAG (DISPLAYNAMEFLAG, pv)) || doing_pinout)
-    _draw_pv_name (pv);
-}
-
-static void
-draw_pin (PinType *pin, bool draw_hole)
-{
-  if (doing_pinout)
-    hid_draw_set_color (Output.fgGC, PCB->PinColor);
-  else
-    set_object_color ((AnyObjectType *)pin,
-                      PCB->WarnColor, PCB->PinSelectedColor,
-                      PCB->ConnectedColor, PCB->FoundColor, PCB->PinColor);
-
-  _draw_pv (pin, draw_hole);
-}
-
-static int
-pin_callback (const BoxType * b, void *cl)
-{
-  draw_pin ((PinType *)b, false);
-  return 1;
-}
-
-static void
-draw_via (PinType *via, bool draw_hole)
-{
-  if (doing_pinout)
-    hid_draw_set_color (Output.fgGC, PCB->ViaColor);
-  else
-    set_object_color ((AnyObjectType *)via,
-                      PCB->WarnColor, PCB->ViaSelectedColor,
-                      PCB->ConnectedColor, PCB->FoundColor, PCB->ViaColor);
-
-  _draw_pv (via, draw_hole);
-}
-
-static int
-via_callback (const BoxType * b, void *cl)
-{
-  draw_via ((PinType *)b, false);
-  return 1;
-}
-
-static void
-draw_pad_name (PadType *pad)
-{
-  BoxType box;
-  bool vert;
-  TextType text;
-
-  if (!pad->Name || !pad->Name[0])
-    text.TextString = EMPTY (pad->Number);
-  else
-    text.TextString = EMPTY (TEST_FLAG (SHOWNUMBERFLAG, PCB) ? pad->Number : pad->Name);
-
-  /* should text be vertical ? */
-  vert = (pad->Point1.X == pad->Point2.X);
-
-  if (vert)
-    {
-      box.X1 = pad->Point1.X                      - pad->Thickness / 2;
-      box.Y1 = MAX (pad->Point1.Y, pad->Point2.Y) + pad->Thickness / 2;
-      box.X1 += Settings.PinoutTextOffsetY;
-      box.Y1 -= Settings.PinoutTextOffsetX;
-    }
-  else
-    {
-      box.X1 = MIN (pad->Point1.X, pad->Point2.X) - pad->Thickness / 2;
-      box.Y1 = pad->Point1.Y                      - pad->Thickness / 2;
-      box.X1 += Settings.PinoutTextOffsetX;
-      box.Y1 += Settings.PinoutTextOffsetY;
-    }
-
-  hid_draw_set_color (Output.fgGC, PCB->PinNameColor);
-
-  text.Flags = NoFlags ();
-  /* Set font height to approx 90% of pin thickness */
-  text.Scale = 90 * pad->Thickness / FONT_CAPHEIGHT;
-  text.X = box.X1;
-  text.Y = box.Y1;
-  text.Direction = vert ? 1 : 0;
-
-  hid_draw_pcb_text (Output.fgGC, &text, 0);
-}
-
-static void
-_draw_pad (hidGC gc, PadType *pad, bool clear, bool mask)
-{
-  if (clear && !mask && pad->Clearance <= 0)
-    return;
-
-  if (TEST_FLAG (THINDRAWFLAG, PCB) ||
-      (clear && TEST_FLAG (THINDRAWPOLYFLAG, PCB)))
-    hid_draw_thin_pcb_pad (gc, pad, clear, mask);
-  else
-    hid_draw_fill_pcb_pad (gc, pad, clear, mask);
-}
-
-static void
-draw_pad (PadType *pad)
-{
-  if (doing_pinout)
-    hid_draw_set_color (Output.fgGC, PCB->PinColor);
-  else
-    set_object_color ((AnyObjectType *)pad, PCB->WarnColor,
-                      PCB->PinSelectedColor, PCB->ConnectedColor, PCB->FoundColor,
-                      FRONT (pad) ? PCB->PinColor : PCB->InvisibleObjectsColor);
-
-  _draw_pad (Output.fgGC, pad, false, false);
-
-  if (doing_pinout || TEST_FLAG (DISPLAYNAMEFLAG, pad))
-    draw_pad_name (pad);
-}
-
 static int
 pad_callback (const BoxType * b, void *cl)
 {
@@ -314,7 +144,13 @@ pad_callback (const BoxType * b, void *cl)
   int *side = cl;
 
   if (ON_SIDE (pad, *side))
-    draw_pad (pad);
+    {
+      set_object_color ((AnyObjectType *)pad, PCB->WarnColor,
+                        PCB->PinSelectedColor, PCB->ConnectedColor, PCB->FoundColor,
+                        FRONT (pad) ? PCB->PinColor : PCB->InvisibleObjectsColor);
+
+      dapi->draw_pad (pad, NULL, NULL);
+    }
   return 1;
 }
 
@@ -356,12 +192,26 @@ draw_element_pins_and_pads (ElementType *element)
   PAD_LOOP (element);
   {
     if (doing_pinout || doing_assy || FRONT (pad) || PCB->InvisibleObjectsOn)
-      draw_pad (pad);
+      {
+        set_object_color ((AnyObjectType *)pad, PCB->WarnColor,
+                          PCB->PinSelectedColor, PCB->ConnectedColor, PCB->FoundColor,
+                          FRONT (pad) ? PCB->PinColor : PCB->InvisibleObjectsColor);
+
+        dapi->draw_pad (pad, NULL, NULL);
+      }
   }
   END_LOOP;
   PIN_LOOP (element);
   {
-    draw_pin (pin, true);
+    set_object_color ((AnyObjectType *)pin, PCB->WarnColor, PCB->PinSelectedColor,
+                      PCB->ConnectedColor, PCB->FoundColor, PCB->PinColor);
+
+    dapi->draw_pin (pin, NULL, NULL);
+
+    set_object_color ((AnyObjectType *)pin, PCB->WarnColor,
+                      PCB->PinSelectedColor, NULL, NULL, Settings.BlackColor);
+
+    dapi->draw_pin_hole (pin, NULL, NULL);
   }
   END_LOOP;
 }
@@ -376,67 +226,6 @@ EMark_callback (const BoxType * b, void *cl)
 }
 
 static int
-hole_callback (const BoxType * b, void *cl)
-{
-  PinType *pv = (PinType *) b;
-  int plated = cl ? *(int *) cl : -1;
-
-  if ((plated == 0 && !TEST_FLAG (HOLEFLAG, pv)) ||
-      (plated == 1 &&  TEST_FLAG (HOLEFLAG, pv)))
-    return 1;
-
-  if (TEST_FLAG (THINDRAWFLAG, PCB))
-    {
-      if (!TEST_FLAG (HOLEFLAG, pv))
-        {
-          hid_draw_set_line_cap (Output.fgGC, Round_Cap);
-          hid_draw_set_line_width (Output.fgGC, 0);
-          hid_draw_arc (Output.fgGC, pv->X, pv->Y,
-                        pv->DrillingHole / 2, pv->DrillingHole / 2, 0, 360);
-        }
-    }
-  else
-    hid_draw_fill_circle (Output.bgGC, pv->X, pv->Y, pv->DrillingHole / 2);
-
-  if (TEST_FLAG (HOLEFLAG, pv))
-    {
-      set_object_color ((AnyObjectType *) pv,
-                        PCB->WarnColor, PCB->ViaSelectedColor,
-                        NULL, NULL, Settings.BlackColor);
-
-      hid_draw_set_line_cap (Output.fgGC, Round_Cap);
-      hid_draw_set_line_width (Output.fgGC, 0);
-      hid_draw_arc (Output.fgGC, pv->X, pv->Y,
-                    pv->DrillingHole / 2, pv->DrillingHole / 2, 0, 360);
-    }
-  return 1;
-}
-
-void
-DrawHoles (bool draw_plated, bool draw_unplated, const BoxType *drawn_area)
-{
-  int plated = -1;
-
-  if ( draw_plated && !draw_unplated) plated = 1;
-  if (!draw_plated &&  draw_unplated) plated = 0;
-
-  r_search (PCB->Data->pin_tree, drawn_area, NULL, hole_callback, &plated);
-  r_search (PCB->Data->via_tree, drawn_area, NULL, hole_callback, &plated);
-}
-
-static int
-line_callback (const BoxType * b, void *cl)
-{
-  LayerType *layer = (LayerType *) cl;
-  LineType *line = (LineType *) b;
-
-  set_layer_object_color (layer, (AnyObjectType *) line);
-  hid_draw_pcb_line (Output.fgGC, line);
-
-  return 1;
-}
-
-static int
 rat_callback (const BoxType * b, void *cl)
 {
   RatType *rat = (RatType *)b;
@@ -444,33 +233,7 @@ rat_callback (const BoxType * b, void *cl)
   set_object_color ((AnyObjectType *) rat, NULL, PCB->RatSelectedColor,
                     PCB->ConnectedColor, PCB->FoundColor, PCB->RatColor);
 
-  if (Settings.RatThickness < 100)
-    rat->Thickness = pixel_slop * Settings.RatThickness;
-  /* rats.c set VIAFLAG if this rat goes to a containing poly: draw a donut */
-  if (TEST_FLAG(VIAFLAG, rat))
-    {
-      int w = rat->Thickness;
-
-      if (TEST_FLAG (THINDRAWFLAG, PCB))
-        hid_draw_set_line_width (Output.fgGC, 0);
-      else
-        hid_draw_set_line_width (Output.fgGC, w);
-      hid_draw_arc (Output.fgGC, rat->Point1.X, rat->Point1.Y, w * 2, w * 2, 0, 360);
-    }
-  else
-    hid_draw_pcb_line (Output.fgGC, (LineType *) rat);
-  return 1;
-}
-
-static int
-arc_callback (const BoxType * b, void *cl)
-{
-  LayerType *layer = (LayerType *) cl;
-  ArcType *arc =  (ArcType *) b;
-
-  set_layer_object_color (layer, (AnyObjectType *) arc);
-  hid_draw_pcb_arc (Output.fgGC, arc);
-
+  dapi->draw_rat (rat, NULL, NULL);
   return 1;
 }
 
@@ -490,12 +253,14 @@ draw_element_package (ElementType *element)
   /* draw lines, arcs, text and pins */
   ELEMENTLINE_LOOP (element);
   {
-    hid_draw_pcb_line (Output.fgGC, line);
+    //hid_draw_pcb_line (Output.fgGC, line);
+    dapi->draw_line (line, NULL, NULL);
   }
   END_LOOP;
   ARC_LOOP (element);
   {
-    hid_draw_pcb_arc (Output.fgGC, arc);
+    //hid_draw_pcb_arc (Output.fgGC, arc);
+    dapi->draw_arc (arc, NULL, NULL);
   }
   END_LOOP;
 }
@@ -574,7 +339,7 @@ DrawEverything (const BoxType *drawn_area)
 	{
 	  r_search (PCB->Data->element_tree, drawn_area, NULL, element_callback, &side);
 	  r_search (PCB->Data->name_tree[NAME_INDEX (PCB)], drawn_area, NULL, name_callback, &side);
-	  DrawLayer (&(PCB->Data->Layer[max_copper_layer + side]), drawn_area);
+	  dapi->draw_layer (&(PCB->Data->Layer[max_copper_layer + side]), drawn_area, NULL);
 	}
       r_search (PCB->Data->pad_tree, drawn_area, NULL, pad_callback, &side);
       hid_draw_end_layer (hid_draw);
@@ -597,20 +362,20 @@ DrawEverything (const BoxType *drawn_area)
 
   /* Draw pins, pads, vias below silk */
   if (hid_draw_is_gui (hid_draw))
-    DrawPPV (SWAP_IDENT ? bottom_group : top_group, drawn_area);
+    dapi->draw_ppv (SWAP_IDENT ? bottom_group : top_group, drawn_area, NULL);
   else
     {
       CountHoles (&plated, &unplated, drawn_area);
 
       if (plated && hid_draw_set_layer (hid_draw, "plated-drill", SL (PDRILL, 0), 0))
         {
-          DrawHoles (true, false, drawn_area);
+          dapi->draw_holes (1, drawn_area, NULL);
           hid_draw_end_layer (hid_draw);
         }
 
       if (unplated && hid_draw_set_layer (hid_draw, "unplated-drill", SL (UDRILL, 0), 0))
         {
-          DrawHoles (false, true, drawn_area);
+          dapi->draw_holes (0, drawn_area, NULL);
           hid_draw_end_layer (hid_draw);
         }
     }
@@ -709,8 +474,7 @@ DrawEMark (ElementType *e, Coord X, Coord Y, bool invisible)
       mark_size = MIN (mark_size, pad0->Thickness / 2);
     }
 
-  hid_draw_set_color (Output.fgGC,
-                      invisible ? PCB->InvisibleMarkColor : PCB->ElementColor);
+  hid_draw_set_color (Output.fgGC, invisible ? PCB->InvisibleMarkColor : PCB->ElementColor);
   hid_draw_set_line_cap (Output.fgGC, Trace_Cap);
   hid_draw_set_line_width (Output.fgGC, 0);
   hid_draw_line (Output.fgGC, X - mark_size, Y, X, Y - mark_size);
@@ -730,82 +494,27 @@ DrawEMark (ElementType *e, Coord X, Coord Y, bool invisible)
     }
 }
 
-/* ---------------------------------------------------------------------------
- * Draws pins pads and vias - Always draws for non-gui HIDs,
- * otherwise drawing depends on PCB->PinOn and PCB->ViaOn
- */
-static void
-DrawPPV (int group, const BoxType *drawn_area)
-{
-  int top_group = GetLayerGroupNumberBySide (TOP_SIDE);
-  int bottom_group = GetLayerGroupNumberBySide (BOTTOM_SIDE);
-  int side;
-
-  if (PCB->PinOn || !hid_draw_is_gui (hid_draw))
-    {
-      /* draw element pins */
-      r_search (PCB->Data->pin_tree, drawn_area, NULL, pin_callback, NULL);
-
-      /* draw element pads */
-      if (group == top_group)
-        {
-          side = TOP_SIDE;
-          r_search (PCB->Data->pad_tree, drawn_area, NULL, pad_callback, &side);
-        }
-
-      if (group == bottom_group)
-        {
-          side = BOTTOM_SIDE;
-          r_search (PCB->Data->pad_tree, drawn_area, NULL, pad_callback, &side);
-        }
-    }
-
-  /* draw vias */
-  if (PCB->ViaOn || !hid_draw_is_gui (hid_draw))
-    {
-      r_search (PCB->Data->via_tree, drawn_area, NULL, via_callback, NULL);
-      r_search (PCB->Data->via_tree, drawn_area, NULL, hole_callback, NULL);
-    }
-  if (PCB->PinOn || doing_assy)
-    r_search (PCB->Data->pin_tree, drawn_area, NULL, hole_callback, NULL);
-}
-
 static int
-clearPin_callback (const BoxType * b, void *cl)
+pin_mask_callback (const BoxType * b, void *cl)
 {
-  PinType *pin = (PinType *) b;
-  if (TEST_FLAG (THINDRAWFLAG, PCB) || TEST_FLAG (THINDRAWPOLYFLAG, PCB))
-    hid_draw_thin_pcb_pv (Output.pmGC, Output.pmGC, pin, false, true);
-  else
-    hid_draw_fill_pcb_pv (Output.pmGC, Output.pmGC, pin, false, true);
+  dapi->draw_pin_mask ((PinType *) b, NULL, NULL);
   return 1;
 }
 
-struct poly_info {
-  const BoxType *drawn_area;
-  LayerType *layer;
-};
-
 static int
-poly_callback (const BoxType * b, void *cl)
+via_mask_callback (const BoxType * b, void *cl)
 {
-  struct poly_info *i = cl;
-  PolygonType *polygon = (PolygonType *)b;
-
-  set_layer_object_color (i->layer, (AnyObjectType *) polygon);
-
-  hid_draw_pcb_polygon (Output.fgGC, polygon, i->drawn_area);
-
+  dapi->draw_via_mask ((PinType *) b, NULL, NULL);
   return 1;
 }
 
 static int
-clearPad_callback (const BoxType * b, void *cl)
+pad_mask_callback (const BoxType * b, void *cl)
 {
   PadType *pad = (PadType *) b;
   int *side = cl;
-  if (ON_SIDE (pad, *side) && pad->Mask)
-    _draw_pad (Output.pmGC, pad, true, true);
+  if (ON_SIDE (pad, *side))
+    dapi->draw_pad_mask (pad, NULL, NULL);
   return 1;
 }
 
@@ -827,7 +536,7 @@ DrawSilk (int side, const BoxType * drawn_area)
     {
       hid_draw_use_mask (hid_draw, HID_MASK_BEFORE);
 #endif
-      DrawLayer (LAYER_PTR (max_copper_layer + side), drawn_area);
+      dapi->draw_layer (LAYER_PTR (max_copper_layer + side), drawn_area, NULL);
       /* draw package */
       r_search (PCB->Data->element_tree, drawn_area, NULL, element_callback, &side);
       r_search (PCB->Data->name_tree[NAME_INDEX (PCB)], drawn_area, NULL, name_callback, &side);
@@ -835,14 +544,14 @@ DrawSilk (int side, const BoxType * drawn_area)
     }
 
   hid_draw_use_mask (hid_draw, HID_MASK_CLEAR);
-  r_search (PCB->Data->pin_tree, drawn_area, NULL, clearPin_callback, NULL);
-  r_search (PCB->Data->via_tree, drawn_area, NULL, clearPin_callback, NULL);
-  r_search (PCB->Data->pad_tree, drawn_area, NULL, clearPad_callback, &side);
+  r_search (PCB->Data->pin_tree, drawn_area, NULL, pin_mask_callback, NULL);
+  r_search (PCB->Data->via_tree, drawn_area, NULL, via_mask_callback, NULL);
+  r_search (PCB->Data->pad_tree, drawn_area, NULL, pad_mask_callback, &side);
 
   if (hid_draw->poly_after)
     {
       hid_draw_use_mask (hid_draw, HID_MASK_AFTER);
-      DrawLayer (LAYER_PTR (max_copper_layer + layer), drawn_area);
+      dapi->draw_layer (LAYER_PTR (max_copper_layer + layer), drawn_area, NULL);
       /* draw package */
       r_search (PCB->Data->element_tree, drawn_area, NULL, element_callback, &side);
       r_search (PCB->Data->name_tree[NAME_INDEX (PCB)], drawn_area, NULL, name_callback, &side);
@@ -869,6 +578,11 @@ DrawMaskBoardArea (int mask_type, const BoxType *drawn_area)
                                      drawn_area->X2, drawn_area->Y2);
 }
 
+struct poly_info {
+  const const BoxType *drawn_area;
+  LayerType *layer;
+};
+
 static int
 mask_poly_callback (const BoxType * b, void *cl)
 {
@@ -938,9 +652,9 @@ DrawMask (int side, const BoxType *screen)
   r_search (Layer->arc_tree,     screen, NULL, mask_arc_callback,  Layer);
   r_search (Layer->text_tree,    screen, NULL, mask_text_callback, Layer);
 
-  r_search (PCB->Data->pin_tree, screen, NULL, clearPin_callback, NULL);
-  r_search (PCB->Data->via_tree, screen, NULL, clearPin_callback, NULL);
-  r_search (PCB->Data->pad_tree, screen, NULL, clearPad_callback, &side);
+  r_search (PCB->Data->pin_tree, screen, NULL, pin_mask_callback, NULL);
+  r_search (PCB->Data->via_tree, screen, NULL, via_mask_callback, NULL);
+  r_search (PCB->Data->pad_tree, screen, NULL, pad_mask_callback, &side);
 
   if (thin)
     hid_draw_set_color (Output.pmGC, "erase");
@@ -960,13 +674,8 @@ DrawPaste (int side, const BoxType *drawn_area)
   hid_draw_set_color (Output.fgGC, PCB->ElementColor);
   ALLPAD_LOOP (PCB->Data);
   {
-    if (ON_SIDE (pad, side) && !TEST_FLAG (NOPASTEFLAG, pad) && pad->Mask > 0)
-      {
-        if (pad->Mask < pad->Thickness)
-          _draw_pad (Output.fgGC, pad, true, true);
-        else
-          _draw_pad (Output.fgGC, pad, false, false);
-      }
+    if (ON_SIDE (pad, side))
+      dapi->draw_pad_paste (pad, NULL, NULL);
   }
   ENDALL_LOOP;
 }
@@ -987,59 +696,6 @@ DrawRats (const BoxType *drawn_area)
     hid_draw_use_mask (hid_draw, HID_MASK_OFF);
 }
 
-static int
-text_callback (const BoxType * b, void *cl)
-{
-  LayerType *layer = cl;
-  TextType *text = (TextType *)b;
-  int min_silk_line;
-
-  if (TEST_FLAG (SELECTEDFLAG, text))
-    hid_draw_set_color (Output.fgGC, layer->SelectedColor);
-  else
-    hid_draw_set_color (Output.fgGC, layer->Color);
-  if (layer == &PCB->Data->SILKLAYER ||
-      layer == &PCB->Data->BACKSILKLAYER)
-    min_silk_line = PCB->minSlk;
-  else
-    min_silk_line = PCB->minWid;
-  hid_draw_pcb_text (Output.fgGC, text, min_silk_line);
-  return 1;
-}
-
-void
-DrawLayer (LayerType *Layer, const BoxType *screen)
-{
-  struct poly_info info = {screen, Layer};
-
-  /* print the non-clearing polys */
-  r_search (Layer->polygon_tree, screen, NULL, poly_callback, &info);
-
-  if (TEST_FLAG (CHECKPLANESFLAG, PCB))
-    return;
-
-  /* draw all visible lines this layer */
-  r_search (Layer->line_tree, screen, NULL, line_callback, Layer);
-
-  /* draw the layer arcs on screen */
-  r_search (Layer->arc_tree, screen, NULL, arc_callback, Layer);
-
-  /* draw the layer text on screen */
-  r_search (Layer->text_tree, screen, NULL, text_callback, Layer);
-
-  /* We should check for hid_draw_is_gui (hid_draw) here, but it's kinda cool seeing the
-     auto-outline magically disappear when you first add something to
-     the "outline" layer.  */
-  if (IsLayerEmpty (Layer)
-      && (strcmp (Layer->Name, "outline") == 0
-	  || strcmp (Layer->Name, "route") == 0))
-    {
-      hid_draw_set_color (Output.fgGC, Layer->Color);
-      hid_draw_set_line_width (Output.fgGC, PCB->minWid);
-      hid_draw_rect (Output.fgGC, 0, 0, PCB->MaxWidth, PCB->MaxHeight);
-    }
-}
-
 /* ---------------------------------------------------------------------------
  * draws one layer group.  If the exporter is not a GUI,
  * also draws the pins / pads / vias in this layer group.
@@ -1061,13 +717,13 @@ DrawLayerGroup (int group, const BoxType *drawn_area)
           strcmp (Layer->Name, "route") == 0)
         rv = 0;
       if (layernum < max_copper_layer && Layer->On)
-        DrawLayer (Layer, drawn_area);
+        dapi->draw_layer (Layer, drawn_area, NULL);
     }
   if (n_entries > 1)
     rv = 1;
 
   if (rv && !hid_draw_is_gui (hid_draw))
-    DrawPPV (group, drawn_area);
+    dapi->draw_ppv (group, drawn_area, NULL);
 }
 
 static void
diff --git a/src/draw.h b/src/draw.h
index b58e5cb..0cb0f88 100644
--- a/src/draw.h
+++ b/src/draw.h
@@ -50,7 +50,6 @@ void DrawElementName (ElementType *);
 void DrawElementPackage (ElementType *);
 void DrawElementPinsAndPads (ElementType *);
 void DrawObject (int, void *, void *);
-void DrawLayer (LayerType *, const BoxType *);
 void EraseVia (PinType *);
 void EraseRat (RatType *);
 void EraseViaName (PinType *);
diff --git a/src/draw_funcs.c b/src/draw_funcs.c
new file mode 100644
index 0000000..8329d49
--- /dev/null
+++ b/src/draw_funcs.c
@@ -0,0 +1,458 @@
+
+#include "global.h"
+#include "data.h"
+#include "misc.h"
+#include "rtree.h"
+#include "draw_funcs.h"
+#include "draw.h"
+#include "hid_draw.h"
+
+static void
+_draw_pv (PinType *pv, bool draw_hole)
+{
+  if (TEST_FLAG (THINDRAWFLAG, PCB))
+    hid_draw_thin_pcb_pv (Output.fgGC, Output.fgGC, pv, draw_hole, false);
+  else
+    hid_draw_fill_pcb_pv (Output.fgGC, Output.bgGC, pv, draw_hole, false);
+}
+
+static void
+draw_pin (PinType *pin, const BoxType *drawn_area, void *userdata)
+{
+  _draw_pv (pin, false);
+}
+
+static void
+draw_pin_mask (PinType *pin, const BoxType *drawn_area, void *userdata)
+{
+  if (TEST_FLAG (THINDRAWFLAG, PCB) || TEST_FLAG (THINDRAWPOLYFLAG, PCB))
+    hid_draw_thin_pcb_pv (Output.pmGC, Output.pmGC, pin, false, true);
+  else
+    hid_draw_fill_pcb_pv (Output.pmGC, Output.pmGC, pin, false, true);
+}
+
+static void
+draw_via (PinType *via, const BoxType *drawn_area, void *userdata)
+{
+  _draw_pv (via, false);
+}
+
+static void
+draw_via_mask (PinType *via, const BoxType *drawn_area, void *userdata)
+{
+  if (TEST_FLAG (THINDRAWFLAG, PCB) || TEST_FLAG (THINDRAWPOLYFLAG, PCB))
+    hid_draw_thin_pcb_pv (Output.pmGC, Output.pmGC, via, false, true);
+  else
+    hid_draw_fill_pcb_pv (Output.pmGC, Output.pmGC, via, false, true);
+}
+
+static void
+draw_hole (PinType *pv, const BoxType *drawn_area, void *userdata)
+{
+  if (!TEST_FLAG (THINDRAWFLAG, PCB))
+    hid_draw_fill_circle (Output.bgGC, pv->X, pv->Y, pv->DrillingHole / 2);
+
+  if (TEST_FLAG (THINDRAWFLAG, PCB) || TEST_FLAG (HOLEFLAG, pv))
+    {
+      hid_draw_set_line_cap (Output.fgGC, Round_Cap);
+      hid_draw_set_line_width (Output.fgGC, 0);
+
+      hid_draw_arc (Output.fgGC, pv->X, pv->Y, pv->DrillingHole / 2, pv->DrillingHole / 2, 0, 360);
+    }
+}
+
+static void
+_draw_pad (hidGC gc, PadType *pad, bool clear, bool mask)
+{
+  if (clear && !mask && pad->Clearance <= 0)
+    return;
+
+  if (TEST_FLAG (THINDRAWFLAG, PCB) ||
+      (clear && TEST_FLAG (THINDRAWPOLYFLAG, PCB)))
+    hid_draw_thin_pcb_pad (gc, pad, clear, mask);
+  else
+    hid_draw_fill_pcb_pad (gc, pad, clear, mask);
+}
+
+static void
+draw_pad (PadType *pad, const BoxType *drawn_area, void *userdata)
+{
+  _draw_pad (Output.fgGC, pad, false, false);
+}
+
+static void
+draw_pad_mask (PadType *pad, const BoxType *drawn_area, void *userdata)
+{
+  if (pad->Mask <= 0)
+    return;
+
+  _draw_pad (Output.pmGC, pad, true, true);
+}
+
+static void
+draw_pad_paste (PadType *pad, const BoxType *drawn_area, void *userdata)
+{
+  if (TEST_FLAG (NOPASTEFLAG, pad) || pad->Mask <= 0)
+    return;
+
+  if (pad->Mask < pad->Thickness)
+    _draw_pad (Output.fgGC, pad, true, true);
+  else
+    _draw_pad (Output.fgGC, pad, false, false);
+}
+
+static void
+draw_line (LineType *line, const BoxType *drawn_area, void *userdata)
+{
+  hid_draw_pcb_line (Output.fgGC, line);
+}
+
+static void
+draw_rat (RatType *rat, const BoxType *drawn_area, void *userdata)
+{
+  if (Settings.RatThickness < 100)
+    rat->Thickness = pixel_slop * Settings.RatThickness;
+  /* rats.c set VIAFLAG if this rat goes to a containing poly: draw a donut */
+  if (TEST_FLAG(VIAFLAG, rat))
+    {
+      int w = rat->Thickness;
+
+      if (TEST_FLAG (THINDRAWFLAG, PCB))
+        hid_draw_set_line_width (Output.fgGC, 0);
+      else
+        hid_draw_set_line_width (Output.fgGC, w);
+      hid_draw_arc (Output.fgGC, rat->Point1.X, rat->Point1.Y, w * 2, w * 2, 0, 360);
+    }
+  else
+    hid_draw_pcb_line (Output.fgGC, (LineType *) rat);
+}
+
+static void
+draw_arc (ArcType *arc, const BoxType *drawn_area, void *userdata)
+{
+  hid_draw_pcb_arc (Output.fgGC, arc);
+}
+
+static void
+draw_poly (PolygonType *polygon, const BoxType *drawn_area, void *userdata)
+{
+  hid_draw_pcb_polygon (Output.fgGC, polygon, drawn_area);
+}
+
+static void
+set_object_color (AnyObjectType *obj, char *warn_color, char *selected_color,
+                  char *connected_color, char *found_color, char *normal_color)
+{
+  char *color;
+
+  if      (warn_color      != NULL && TEST_FLAG (WARNFLAG,      obj)) color = warn_color;
+  else if (selected_color  != NULL && TEST_FLAG (SELECTEDFLAG,  obj)) color = selected_color;
+  else if (connected_color != NULL && TEST_FLAG (CONNECTEDFLAG, obj)) color = connected_color;
+  else if (found_color     != NULL && TEST_FLAG (FOUNDFLAG,     obj)) color = found_color;
+  else                                                                color = normal_color;
+
+  hid_draw_set_color (Output.fgGC, color);
+}
+
+static void
+set_layer_object_color (LayerType *layer, AnyObjectType *obj)
+{
+  set_object_color (obj, NULL, layer->SelectedColor, PCB->ConnectedColor, PCB->FoundColor, layer->Color);
+}
+
+static int
+line_callback (const BoxType * b, void *cl)
+{
+  LayerType *layer = cl;
+  LineType *line = (LineType *)b;
+
+  set_layer_object_color (layer, (AnyObjectType *) line);
+  dapi->draw_line (line, NULL, NULL);
+  return 1;
+}
+
+static int
+arc_callback (const BoxType * b, void *cl)
+{
+  LayerType *layer = cl;
+  ArcType *arc = (ArcType *)b;
+
+  set_layer_object_color (layer, (AnyObjectType *) arc);
+  dapi->draw_arc (arc, NULL, NULL);
+  return 1;
+}
+
+struct poly_info {
+  const const BoxType *drawn_area;
+  LayerType *layer;
+};
+
+static int
+poly_callback (const BoxType * b, void *cl)
+{
+  struct poly_info *i = cl;
+  PolygonType *polygon = (PolygonType *)b;
+
+  set_layer_object_color (i->layer, (AnyObjectType *) polygon);
+  dapi->draw_poly (polygon, i->drawn_area, NULL);
+  return 1;
+}
+
+static int
+text_callback (const BoxType * b, void *cl)
+{
+  LayerType *layer = cl;
+  TextType *text = (TextType *)b;
+  int min_silk_line;
+
+  if (TEST_FLAG (SELECTEDFLAG, text))
+    hid_draw_set_color (Output.fgGC, layer->SelectedColor);
+  else
+    hid_draw_set_color (Output.fgGC, layer->Color);
+  if (layer == &PCB->Data->SILKLAYER ||
+      layer == &PCB->Data->BACKSILKLAYER)
+    min_silk_line = PCB->minSlk;
+  else
+    min_silk_line = PCB->minWid;
+  hid_draw_pcb_text (Output.fgGC, text, min_silk_line);
+  return 1;
+}
+
+static void
+set_pv_inlayer_color (PinType *pv, LayerType *layer, int type)
+{
+  if (TEST_FLAG (WARNFLAG, pv))          hid_draw_set_color (Output.fgGC, PCB->WarnColor);
+  else if (TEST_FLAG (SELECTEDFLAG, pv)) hid_draw_set_color (Output.fgGC, (type == VIA_TYPE) ? PCB->ViaSelectedColor
+                                                                                             : PCB->PinSelectedColor);
+  else if (TEST_FLAG (FOUNDFLAG, pv))    hid_draw_set_color (Output.fgGC, PCB->ConnectedColor);
+  else                                   hid_draw_set_color (Output.fgGC, layer->Color);
+}
+
+static int
+pin_inlayer_callback (const BoxType * b, void *cl)
+{
+  set_pv_inlayer_color ((PinType *)b, cl, PIN_TYPE);
+  dapi->draw_pin ((PinType *)b, NULL, NULL);
+  return 1;
+}
+
+static int
+via_inlayer_callback (const BoxType * b, void *cl)
+{
+  set_pv_inlayer_color ((PinType *)b, cl, VIA_TYPE);
+  dapi->draw_via ((PinType *)b, NULL, NULL);
+  return 1;
+}
+
+static int
+pad_inlayer_callback (const BoxType * b, void *cl)
+{
+  PadType *pad = (PadType *)b;
+  LayerType *layer = cl;
+  int bottom_group = GetLayerGroupNumberBySide (BOTTOM_SIDE);
+  int group = GetLayerGroupNumberByPointer (layer);
+
+  int side = (group == bottom_group) ? BOTTOM_SIDE : TOP_SIDE;
+
+  if (ON_SIDE (pad, side))
+    {
+      if (TEST_FLAG (WARNFLAG, pad))          hid_draw_set_color (Output.fgGC, PCB->WarnColor);
+      else if (TEST_FLAG (SELECTEDFLAG, pad)) hid_draw_set_color (Output.fgGC, PCB->PinSelectedColor);
+      else if (TEST_FLAG (FOUNDFLAG, pad))    hid_draw_set_color (Output.fgGC, PCB->ConnectedColor);
+      else                                    hid_draw_set_color (Output.fgGC, layer->Color);
+
+      dapi->draw_pad (pad, NULL, NULL);
+    }
+  return 1;
+}
+
+static int
+pin_hole_callback (const BoxType * b, void *cl)
+{
+  PinType *pin = (PinType *)b;
+  int plated = cl ? *(int *) cl : -1;
+
+  if ((plated == 0 && !TEST_FLAG (HOLEFLAG, pin)) ||
+      (plated == 1 &&  TEST_FLAG (HOLEFLAG, pin)))
+    return 1;
+
+  set_object_color ((AnyObjectType *) pin, PCB->WarnColor,
+                    PCB->PinSelectedColor, NULL, NULL, Settings.BlackColor);
+
+  dapi->draw_pin_hole (pin, NULL, NULL);
+  return 1;
+}
+
+static int
+via_hole_callback (const BoxType * b, void *cl)
+{
+  PinType *via = (PinType *)b;
+  int plated = cl ? *(int *) cl : -1;
+
+  if ((plated == 0 && !TEST_FLAG (HOLEFLAG, via)) ||
+      (plated == 1 &&  TEST_FLAG (HOLEFLAG, via)))
+    return 1;
+
+  set_object_color ((AnyObjectType *) via, PCB->WarnColor,
+                    PCB->ViaSelectedColor, NULL, NULL, Settings.BlackColor);
+
+  dapi->draw_via_hole (via, NULL, NULL);
+  return 1;
+}
+
+static int
+pin_callback (const BoxType * b, void *cl)
+{
+  set_object_color ((AnyObjectType *)b,
+                    PCB->WarnColor, PCB->PinSelectedColor,
+                    PCB->ConnectedColor, PCB->FoundColor, PCB->PinColor);
+
+  dapi->draw_pin ((PinType *)b, NULL, NULL);
+  return 1;
+}
+
+static int
+via_callback (const BoxType * b, void *cl)
+{
+  set_object_color ((AnyObjectType *)b,
+                    PCB->WarnColor, PCB->ViaSelectedColor,
+                    PCB->ConnectedColor, PCB->FoundColor, PCB->ViaColor);
+
+  dapi->draw_via ((PinType *)b, NULL, NULL);
+  return 1;
+}
+
+static int
+pad_callback (const BoxType * b, void *cl)
+{
+  PadType *pad = (PadType *) b;
+  int *side = cl;
+
+  if (ON_SIDE (pad, *side))
+    {
+      set_object_color ((AnyObjectType *)pad, PCB->WarnColor,
+                        PCB->PinSelectedColor, PCB->ConnectedColor, PCB->FoundColor,
+                        FRONT (pad) ? PCB->PinColor : PCB->InvisibleObjectsColor);
+
+      dapi->draw_pad (pad, NULL, NULL);
+    }
+  return 1;
+}
+
+static void
+draw_ppv (int group, const BoxType *drawn_area, void *userdata)
+{
+  int top_group = GetLayerGroupNumberBySide (TOP_SIDE);
+  int bottom_group = GetLayerGroupNumberBySide (BOTTOM_SIDE);
+  int side;
+
+  if (PCB->PinOn || !gui->gui)
+    {
+      /* draw element pins */
+      r_search (PCB->Data->pin_tree, drawn_area, NULL, pin_callback, NULL);
+
+      /* draw element pads */
+      if (group == top_group)
+        {
+          side = TOP_SIDE;
+          r_search (PCB->Data->pad_tree, drawn_area, NULL, pad_callback, &side);
+        }
+
+      if (group == bottom_group)
+        {
+          side = BOTTOM_SIDE;
+          r_search (PCB->Data->pad_tree, drawn_area, NULL, pad_callback, &side);
+        }
+    }
+
+  /* draw vias */
+  if (PCB->ViaOn)
+    r_search (PCB->Data->via_tree, drawn_area, NULL, via_callback, NULL);
+
+  dapi->draw_holes (-1, drawn_area, NULL);
+}
+
+static void
+draw_holes (int plated, const BoxType *drawn_area, void *userdata)
+{
+  if (PCB->PinOn)
+    r_search (PCB->Data->pin_tree, drawn_area, NULL, pin_hole_callback, &plated);
+  if (PCB->ViaOn)
+    r_search (PCB->Data->via_tree, drawn_area, NULL, via_hole_callback, &plated);
+}
+
+static void
+draw_layer (LayerType *layer, const BoxType *drawn_area, void *userdata)
+{
+  int top_group = GetLayerGroupNumberBySide (TOP_SIDE);
+  int bottom_group = GetLayerGroupNumberBySide (BOTTOM_SIDE);
+  int layer_num = GetLayerNumber (PCB->Data, layer);
+  int group = GetLayerGroupNumberByPointer (layer);
+  struct poly_info info = {drawn_area, layer};
+
+  /* print the non-clearing polys */
+  r_search (layer->polygon_tree, drawn_area, NULL, poly_callback, &info);
+
+  if (TEST_FLAG (CHECKPLANESFLAG, PCB))
+    return;
+
+  r_search (layer->line_tree, drawn_area, NULL, line_callback, layer);
+  r_search (layer->arc_tree,  drawn_area, NULL, arc_callback,  layer);
+  r_search (layer->text_tree, drawn_area, NULL, text_callback, layer);
+
+  /* We should check for gui->gui here, but it's kinda cool seeing the
+     auto-outline magically disappear when you first add something to
+     the "outline" layer.  */
+
+  if (strcmp (layer->Name, "outline") == 0 ||
+      strcmp (layer->Name, "route") == 0)
+    {
+      if (IsLayerEmpty (layer))
+        {
+          hid_draw_set_color (Output.fgGC, layer->Color);
+          hid_draw_set_line_width (Output.fgGC, PCB->minWid);
+          hid_draw_rect (Output.fgGC, 0, 0, PCB->MaxWidth, PCB->MaxHeight);
+        }
+      return;
+    }
+
+  /* Don't draw vias on silk layers */
+  if (layer_num >= max_copper_layer)
+    return;
+
+  r_search (PCB->Data->pin_tree, drawn_area, NULL, pin_inlayer_callback, layer);
+
+  /* draw element pads */
+  if (group == top_group)
+    r_search (PCB->Data->pad_tree, drawn_area, NULL, pad_inlayer_callback, layer);
+
+  if (group == bottom_group)
+    r_search (PCB->Data->pad_tree, drawn_area, NULL, pad_inlayer_callback, layer);
+
+  /* draw vias */
+  r_search (PCB->Data->via_tree, drawn_area, NULL, via_inlayer_callback, layer);
+  r_search (PCB->Data->pin_tree, drawn_area, NULL, pin_hole_callback, NULL);
+  r_search (PCB->Data->via_tree, drawn_area, NULL, via_hole_callback, NULL);
+}
+
+struct draw_funcs d_f = {
+  .draw_pin       = draw_pin,
+  .draw_pin_mask  = draw_pin_mask,
+  .draw_pin_hole  = draw_hole,
+  .draw_via       = draw_via,
+  .draw_via_mask  = draw_via_mask,
+  .draw_via_hole  = draw_hole,
+  .draw_pad       = draw_pad,
+  .draw_pad_mask  = draw_pad_mask,
+  .draw_pad_paste = draw_pad_paste,
+  .draw_line      = draw_line,
+  .draw_rat       = draw_rat,
+  .draw_arc       = draw_arc,
+  .draw_poly      = draw_poly,
+
+  .draw_ppv       = draw_ppv,
+  .draw_holes     = draw_holes,
+  .draw_layer     = draw_layer,
+};
+
+struct draw_funcs *dapi = &d_f;
diff --git a/src/draw_funcs.h b/src/draw_funcs.h
new file mode 100644
index 0000000..ba4982d
--- /dev/null
+++ b/src/draw_funcs.h
@@ -0,0 +1,20 @@
+struct draw_funcs {
+  void (*draw_pin)       (PinType *,     const BoxType *, void *);
+  void (*draw_pin_mask)  (PinType *,     const BoxType *, void *);
+  void (*draw_pin_hole)  (PinType *,     const BoxType *, void *);
+  void (*draw_via)       (PinType *,     const BoxType *, void *);
+  void (*draw_via_mask)  (PinType *,     const BoxType *, void *);
+  void (*draw_via_hole)  (PinType *,     const BoxType *, void *);
+  void (*draw_pad)       (PadType *,     const BoxType *, void *);
+  void (*draw_pad_mask)  (PadType *,     const BoxType *, void *);
+  void (*draw_pad_paste) (PadType *,     const BoxType *, void *);
+  void (*draw_line)      (LineType *,    const BoxType *, void *);
+  void (*draw_rat)       (RatType *,     const BoxType *, void *);
+  void (*draw_arc)       (ArcType *,     const BoxType *, void *);
+  void (*draw_poly)      (PolygonType *, const BoxType *, void *);
+  void (*draw_ppv)       (int,           const BoxType *, void *);
+  void (*draw_holes)     (int,           const BoxType *, void *);
+  void (*draw_layer)     (LayerType *,   const BoxType *, void *);
+};
+
+extern struct draw_funcs *dapi;
diff --git a/src/hid/gerber/gerber.c b/src/hid/gerber/gerber.c
index dc80eff..519264e 100644
--- a/src/hid/gerber/gerber.c
+++ b/src/hid/gerber/gerber.c
@@ -25,6 +25,7 @@
 #include "error.h"
 #include "draw.h"
 #include "pcb-printf.h"
+#include "draw_funcs.h"
 
 #include "hid.h"
 #include "hid_draw.h"
@@ -913,7 +914,7 @@ gerber_set_layer (const char *name, int group, int empty)
     {
       if (outline_layer
 	  && outline_layer != PCB->Data->Layer+idx)
-	DrawLayer (outline_layer, &region);
+	dapi->draw_layer (outline_layer, &region, NULL);
       else if (!outline_layer)
 	{
 	  hidGC gc = hid_draw_make_gc (&gerber_graphics);
diff --git a/src/hid/ps/ps.c b/src/hid/ps/ps.c
index e671ff3..35b287a 100644
--- a/src/hid/ps/ps.c
+++ b/src/hid/ps/ps.c
@@ -15,6 +15,7 @@
 #include "error.h"
 #include "draw.h"
 #include "pcb-printf.h"
+#include "draw_funcs.h"
 
 #include "hid.h"
 #include "hid_draw.h"
@@ -1007,7 +1008,7 @@ ps_set_layer (const char *name, int group, int empty)
       strcmp (name, "route") != 0
       )
     {
-      DrawLayer (global.outline_layer, &global.region);
+      dapi->draw_layer (global.outline_layer, &global.region, NULL);
     }
 
   return 1;
